### ISSUE TITLE
[RFC] Device Placement API for RPC 

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#41369]
## Goal

As of v1.6, `torch.distributed.rpc` only accepts CPU tensors. Applications have to first move the tensor to CPU and then pass it to RPC APIs. This is both inconvenient and slow. The original reason for enforcing CPU tensors is to avoid invalid devices on the destination process. The RPC device placement API aims at (partly) addressing this issue by allowing applications to configure the device mapping on caller and callee. Note that this proposal focuses on changes in API and RPC agents. There will be separate efforts in the comm layer and the distributed autograd engine to improve its efficiency. 

## API

We can add one `map_location` argument to [`RpcBackendOptions`](https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.RpcBackendOptions). The example below shows how to configure this field. 

```python
rpc.init_rpc(
    "worker0",
    rank=0,
    world_size=2,
    backend=rpc.BackendType.TENSORPIPE,
    rpc_backend_options=rpc.TensorPipeRpcBackendOptions(
        num_worker_threads=8,
        rpc_timeout=20, # 20 second timeout
        map_location={
        	"worker1": {
        		"cpu":     	"cuda:0",
        		"cuda:0":  	"cuda:1"
        	},
        	"worker2": {
        		"cpu": 		"cuda:1",
        		"cuda:0":	"cuda:2"
        	}
        }
    )
)

x = torch.zeros(2).to(0)
y = torch.ones(2).to(0)

# x and y will be materialized on "cuda:1" in "worker1".
# the return value z will be materialized on "cuda:0" on "worker0",
# which is defined by the inverse of the mapping specified by "worker0"
z = rpc.rpc_sync("worker1", torch.add, args=(x, y))

# x and y will be materialized on "cuda:2" in "worker2".
# the return value z will be materialized on "cuda:0" on "worker0".
z = rpc.rpc_sync("worker2", torch.add, args=(x, y))

rpc.shutdown()
```

In this proposal, we only aim to add default `map_location`. If necessary, we can add the per-RPC `map_location` later. 

This is only intended to be added to the `TensorPipe` backend. `ProcessGroup` backend will throw an exception is this field is specified.

### Device Placement for Return Value

There are a few options here.

* Option 1: we can use the inverse of the `map_location` to place tensors in the return value.
* Option 2: we can add another `return_map_location` for this purpose. 
* Option 3: use the `map_location` defined on the destination to send return values. 

The downside of Option 3 is that the caller cannot fully control the behavior of its own RPC. Imagine a Parameter-Server
training application with one PS with multiple trainers. Option 3 would requre the PS to define the `map_location` entry
for all trainers, which is cumbersome. So it might be more convenient for applications to use Option 1 or 2.  

As the first version, we can go with Option 1, and when we see requests for Option 2, we can add a `return_map_location` and
use entries defined there to override the inverse of `map_location`, so that there will be no BC issue.

## Implementation

In the `init_rpc` call, all processes will use the initialized RPC to communicate `map_location` to make sure that there 
are no invalid devices. If there is any in any process, all processes will throw an error. 

This `map_location` will be stored in the agent and not necessary to modify `Message` structure. 

As there is no remote device type yet, agent will first use integers to store devices, -1 for CPU.

**Comment by rohan-varma** [comment]
For Option 1, what happens if the mapping is not invertible? 

"worker1": {
        		"cpu":     	"cuda:1",
        		"cuda:0":  	"cuda:1"
        	}

For a return value being sent to worker 1, it seems like we won't be able choose whether to materialize it on cpu or cuda:0. Is the plan to have the framework "remember" the mapping that was used for the particular RPC, and use that to help with the inversion?

**Comment by mrshenli** [comment]
> For Option 1, what happens if the mapping is not invertible?

It can throw an error on all processes when calling `init_rpc`. And if this is an issue, the application can specify the `return_map_location` arg as well to override the return mapping. 

>  Is the plan to have the framework "remember" the mapping that was used for the particular RPC, and use that to help with the inversion?

For default device placement, we can let all agents remember all egress mappings at construction time and consult this mapping for all RPCs. If we need to add per-RPC device placement, then yes, the Message will need to carry the mapping for the return value.

**Comment by jjlilley** [comment]
Wondering (maybe this isn't a good idea), would we consider allowing the caller to explicitly specify device placement, e.g.
``` 
   a = rpc.rpc_sync("worker0:0", torch.add, args=(x, y))
   b = rpc.rpc_sync("worker0:1", torch.add, args=(x, y))
```

**Comment by mrshenli** [comment]
> Wondering (maybe this isn't a good idea), would we consider allowing the caller to explicitly specify device placement, e.g.

Yep, this would work if `x` and `y` will be placed on the same device on the callee, but if they will be allocated to different devices (say we are using a custom target function instead of `torch.add`), then we will need more than one number after `worker0:`.

**Comment by lw** [comment]
In principle I'm all up for putting the default mapping in the backend opts: it makes it specific to the TP agent, it makes it immutable which simplifies things, it matches what we do for timeouts (whose path we might follow for the mapping: first a global option then, at a later time, a per-call override). But now that I see it written down I realize that it requires to give, all at once, a mapping for each worker. This looks rather cumbersome, and it reminds me of the `worker_name_to_id` map we had which we tried to hard to get rid of (#37784).

I'm wondering if we could do it another way, for example through a global function on the module, like `torch.distributed.rpc.set_device_mapping("worker0", {"cuda:0": "cuda:1"})`. This means the mapping wouldn't be immutable, and in fact we would already allow a different mapping for each RPC call (by changing it just before the call), which means it's going to be quite harder to implement.

Another thing to note is that TensorPipe doesn't plan to support device-to-host transfers (and vice versa) at this time, so there is no point in adding the CPU among the devices, since it can only be mapped to/from the CPU.

As a guideline, I think for the very initial implementation we should make the API as strict as possible, as long as it supports some use cases, explicitly rejecting invalid args, so that we can generalize it later. In that spirit, I think requiring the map to be invertible and not having a `return_map_location` sound good approaches.

The use cases we should probably focus on first are model/pipeline parallelism, with multiple workers with one GPU each, plus possibly one leader with multiple GPUs that sends data to these workers.

Using a syntax like "worker0:0" would be equivalent to a mapping where all devices are mapped to "cuda:0" right? If for `map_location` we're following the syntax of `torch.load` (see [here](https://pytorch.org/docs/master/generated/torch.load.html)) then we could achieve this with `map_location="cuda:0"`. This may cause issues with JIT typing though. Also, my feeling is that with "worker0:0" we're "cramming" too much information in a string, giving up on some expressiveness that we could achieve by using a dedicated argument (which doesn't add much overhead for the user).

**Comment by mrshenli** [comment]
> I'm wondering if we could do it another way, for example through a global function on the module, like torch.distributed.rpc.set_device_mapping("worker0", {"cuda:0": "cuda:1"}). This means the mapping wouldn't be immutable, and in fact we would already allow a different mapping for each RPC call (by changing it just before the call), which means it's going to be quite harder to implement.

This makes sense to me and +1 for its flexibility for per-RPC device mapping. One caveat is that, if we do this, we can no longer keep the mapping in the agent, and there has to be a per-message field, as the serialization can occur concurrently on a different thread. But this should be fine, as eventually we will need per-message device mapping field anyway.

**Comment by mrshenli** [comment]
Had a discussion with @lw and @beauby on the API. 

* As @lw mentioned above, let `map_location` take a nested map looks very cumbersome and hard to use. 
* Having an dedicated `torch.distributed.rpc.set_device_mapping` API might require additional synchronization code. Otherwise, it is unclear how to handle nested RPC requests if then arrive before all `torch.distributed.rpc.set_device_mapping` are done. 

An alternative is to expose the `set_map_location` API on `TensorPipeRpcBackendOptions`, sth like:

```python
options = rpc.TensorPipeRpcBackendOptions(num_worker_threads=8)

for w in workers:
    options.set_map_location(w, , {"cuda:0": "cuda:1"})

rpc.init_rpc(
    "worker0",
    rank=0,
    world_size=2,
    backend=rpc.BackendType.TENSORPIPE,
    rpc_backend_options=options
)

In this way, we can all `_wait_all_workers` in `init_rpc` to make sure no nested RPC occurs before all `map_location` are set properly.
